sdks/python/apache_beam/metrics/execution.py

```diff
@@ -122,6 +122,13 @@ def __repr__(self):
     return 'MetricResult(key={}, committed={}, attempted={})'.format(
         self.key, str(self.committed), str(self.attempted))
 
+  @property
+  def result(self):
+    """Short-hand for falling back to attempted metrics if it seems that
+    committed was not populated (e.g. due to not being supported on a given
+    runner"""
+    return self.committed if self.committed else self.attempted
```

Comment:
Stupid question: What is the difference between the two?

Reply:
It is a good question!

I believe "committed" consists only of updates from task-attempts that succeeded, whereas "attempted" contains cumulative updates from all task-attempts; [this comment in MetricResult is somewhat canonical](https://github.com/apache/beam/blob/bb49247808af309a529ab2443b8d779c82970bcb/sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricResult.java#L37-L46). I think [this is the original design doc about it](https://docs.google.com/document/d/1voyUIQ2DrWkoY-BsJwM8YvF4gGKB76CDG8BYL8XBc7A/edit#heading=h.m3vu8fd66s5y), though I think it's out of date now.

Flink only supports "attempted" metrics (in Beam; maybe in general?); [here's an example where it only fetches "attempted" metric updates](https://github.com/apache/beam/blob/bb49247808af309a529ab2443b8d779c82970bcb/runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java#L145). If you have thoughts about the necessity of this restriction from a Flink PoV I'd love to hear them!

A confusing thing is that [`MetricContainerImpl` discusses "commit"s to `MetricCell`s](https://github.com/apache/beam/blob/bb49247808af309a529ab2443b8d779c82970bcb/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java#L39) but I don't think it's referring to the same thing.

Reply:
Thanks for clarifying. Flink does not support committed metrics per se, though it looks like you could work around this by changing the metrics template to include the execution attempt id.